ports/raspberrypi/common-hal/busio/UART.c

```diff
@@ -24,238 +24,175 @@
  * THE SOFTWARE.
  */
 
-#include "shared-bindings/microcontroller/__init__.h"
 #include "shared-bindings/busio/UART.h"
 
-#include "mpconfigport.h"
-#include "lib/utils/interrupt_char.h"
-#include "py/gc.h"
+#include "py/stream.h"
 #include "py/mperrno.h"
 #include "py/runtime.h"
-#include "py/stream.h"
-#include "supervisor/shared/translate.h"
 #include "supervisor/shared/tick.h"
+#include "lib/utils/interrupt_char.h"
+#include "common-hal/microcontroller/Pin.h"
 
-#define UART_DEBUG(...) (void)0
-// #define UART_DEBUG(...) mp_printf(&mp_plat_print __VA_OPT__(,) __VA_ARGS__)
-
-// Do-nothing callback needed so that usart_async code will enable rx interrupts.
-// See comment below re usart_async_register_callback()
-// static void usart_async_rxc_callback(const struct usart_async_descriptor *const descr) {
-//     // Nothing needs to be done by us.
-// }
+#include "src/rp2_common/hardware_irq/include/hardware/irq.h"
+#include "src/rp2_common/hardware_gpio/include/hardware/gpio.h"
 
 #define NO_PIN 0xff
 
+#define UART_INST(uart) (((uart) ? uart1 : uart0))
+
+typedef enum {
+    STATUS_FREE = 0,
+    STATUS_IN_USE,
+    STATUS_NEVER_RESET
+} uart_status_t;
+
+static uart_status_t uart_status[2];
+
+void reset_uart(void) {
+    for (uint8_t num = 0; num < 2; num++) {
+        if (uart_status[num] == STATUS_IN_USE) {
+            uart_status[num] = STATUS_FREE;
+            uart_deinit(UART_INST(num));
+        }
+    }
+}
+
+void never_reset_uart(uint8_t num) {
+    uart_status[num] = STATUS_NEVER_RESET;
+}
+
+static uint8_t get_free_uart() {
+    uint8_t num;
+    for (num = 0; num < 2; num++) {
+        if (uart_status[num] == STATUS_FREE) {
+            break;
+        }
+        if (num) {
+            mp_raise_RuntimeError(translate("All UART peripherals are in use"));
+        }
+    }
+    return num;
+}
+
+static uint8_t pin_init(const uint8_t uart, const mcu_pin_obj_t * pin, const uint8_t pin_type) {
+    if (pin == NULL) {
+        return NO_PIN;
+    }
+    if (!(((pin->number & 3) == pin_type) && ((((pin->number + 4) & 8) >> 3) == uart))) {
+        mp_raise_ValueError(translate("Invalid pins"));
+    }
+    claim_pin(pin);
+    gpio_set_function(pin->number, GPIO_FUNC_UART);
+    return pin->number;
+}
+
+static ringbuf_t ringbuf[2];
+
+static void uart0_callback(void) {
+    while (uart_is_readable(uart0) && !ringbuf_put(&ringbuf[0], (uint8_t)uart_get_hw(uart0)->dr)) {}
+}
+
+static void uart1_callback(void) {
+    while (uart_is_readable(uart1) && !ringbuf_put(&ringbuf[1], (uint8_t)uart_get_hw(uart1)->dr)) {}
```

Comment:
Check `ringbuf_num_empty() > 1` before reading `dr`. Otherwise you risk reading a value you can't store. Some UARTs can use CTS and RTS for control flow so NOT reading is the right thing to do when full.